Clarify rules for which commit is used in CI (#88425)
pytorch/pytorch: CONTRIBUTING.md

### Which commit is used in CI?

For CI run on `master`, this repository is checked out for a given `master`
commit, and CI is run on that commit (there isn't really any other choice).

For PRs, however, it's a bit more complicated. Consider this commit graph, where
`master` is at commit `A`, and the branch for PR #42 (just a placeholder) is at
commit `B`:

```

>        o---o---B (pr-commit)

Reviewer: This one is still good
```suggestion
       o---o---B (refs/pull/42/head)
```

      /         \

>      /           C (merged-commit)

Reviewer: This one is good
```suggestion
     /           C (refs/pull/42/merge)
```

    /           /

> ---o---o---o---A (merge-destination) - usually master

Reviewer: Do you know if `refs/pull/42/base` or `refs/pull/42/target`exists?

Author: Running `git ls-remote` shows that something similar to `refs/pull/42/base` does exist, but it doesn't show any `refs/pull/42/target`

Thought that may be cuz the target is usually `master`

```

There are two possible choices for which commit to use:

1. Checkout commit `B`, the head of the PR (manually committed by the PR
   author).
2. Checkout commit `C`, the hypothetical result of what would happen if the PR
   were merged into it's destination (usually `master`).

For all practical purposes, most people can think of the commit being used as
commit `B` (choice **1**).

> However, if you're modifying the workflow files (which govern CI behavior) there's

Reviewer: ```suggestion
However, if workflow files (which govern CI behavior) were modified (either by your PR or since dev branch were created ) there's
```

a nuance to know about:
The workflow files themselves get taken from checkpoint `C`, the merger of your
PR and the `master` branch. But only the workflow files get taken from that merged
checkpoint. Everything else (tests, code, etc) all get taken directly from your

> PR's commit (commit `B`)

Reviewer: ```suggestion
PR's commit (commit `B`). Please note, this scenario would never affect PRs authored by `ghstack` as they would not automatically ingest the updates from default branch.
```

2. Checkout commit `C`, the hypothetical result of what would happen if the PR
   were merged into `master` (automatically generated by GitHub).

For all practical purposes, most people can think of the commit being used as 

> commit `B` (choice **1**).

Reviewer: This is wrong for the stacked PRs or for PR that do not target master. I.e. A is not `refs/head/master` but commit `refs/pull/42/base` belongs to I guess

Author: I've reworded the chart a bit to use more generic terms for the source commit and the branch the PR would be merged into.

Does this make it more accurate?
